src/annoymodule.cc

```diff
@@ -14,87 +14,490 @@
+    return this->load(filename);
   }
+
+};
+
+
+class AI : public AnnoyIndexPython <int32_t, float, Angular<int32_t, float> > {
+  public:
+  AI(int f) : AnnoyIndexPython <int32_t, float, Angular<int32_t, float> >(f){};
+};
+class EI : public AnnoyIndexPython <int32_t, float, Euclidean<int32_t, float> > {
+  public:
+  EI(int f) : AnnoyIndexPython <int32_t, float, Euclidean<int32_t, float> >(f){};
 };
 
-template<typename C>
-void expose_methods(python::class_<C> c) {
-  c.def("add_item",          &C::add_item_py)
-    .def("build",             &C::build)
-    .def("save",              &C::save_py)
-    .def("load",              &C::load_py)
-    .def("unload",            &C::unload)
-    .def("get_distance",      &C::get_distance)
-    .def("get_nns_by_item",   &C::get_nns_by_item_py)
-    .def("get_nns_by_vector", &C::get_nns_by_vector_py)
-    .def("get_item_vector",   &C::get_item_vector_py)
-    .def("get_n_items",       &C::get_n_items)
-    .def("verbose",           &C::verbose);
+
+// annoy python object
+typedef struct {
+  PyObject_HEAD
+  int f;
+  char ch_metric;
+  void* ptr;
+} py_annoy;
+
+
+static PyObject *
+py_an_new(PyTypeObject *type, PyObject *args, PyObject *kwds) {
+  py_annoy *self;
+
+  self = (py_annoy *)type->tp_alloc(type, 0);
+  if (self != NULL) {
+    self->f = 0;
+    self->ch_metric = 0;
+    self->ptr = NULL;
+  }
+
+  return (PyObject *)self;
+}
+
+
+static int 
+py_an_init(py_annoy *self, PyObject *args, PyObject *kwds) {
+  const char *metric;
+
+  if (!PyArg_ParseTuple(args, "is", &self->f, &metric))
+    return -1;
+  self->ch_metric = metric[0];
+  AI *ai; EI *ei;
+  switch(self->ch_metric) {
+  case 'a':
+    ai = new AI(self->f);
+    self->ptr = static_cast<AI*>(static_cast<void*>(ai));
+    break;
+  case 'e':
+    ei = new EI(self->f);
+    self->ptr = static_cast<EI*>(static_cast<void*>(ei));
+    break;
+  }
+  return 0;
+}
+
+
+static void 
+py_an_dealloc(py_annoy* self) {
+  if (self->ptr) {
+    switch(self->ch_metric) {
+    case 'a':
+      delete (AI*)self->ptr;
+      break;
+    case 'e':
+      delete (EI*)self->ptr;
+      break;
+    }
+  }
+  self->ob_type->tp_free((PyObject*)self);
+}
+
+
+static PyMemberDef py_annoy_members[] = {
+  {(char*)"_f", T_INT, offsetof(py_annoy, f), 0,
+   (char*)""},
+  {(char*)"ch_metric", T_CHAR, offsetof(py_annoy, ch_metric), 0,
+   (char*)"a|e for angular|euclidean"},
+  {NULL}	/* Sentinel */
+};
+
+
+static PyObject *
+py_an_load(py_annoy *self, PyObject *args) {
+  char* filename;
+  bool res = false;
+  if (!self->ptr) 
+    return Py_None;
+  if (!PyArg_ParseTuple(args, "s", &filename))
+    return Py_None;
+  AI *ai; EI *ei;
+  switch(self->ch_metric) {
+  case 'a':
+    ai = static_cast<AI*>(self->ptr);
+    res = ai->load_py(filename);
+    break;
+  case 'e':
+    ei = static_cast<EI*>(self->ptr);
+    res = ei->load_py(filename);
+    break;
+  }
+  if (!res) {
+    PyErr_SetString(PyExc_IOError,"load error");
```

Comment:
It should be safe to use PyErr_SetFromErrno(PyExc_IOError); here, as errno should be preserved from the load_py call. I think.